functional-programming-spreadsheet/index890.js

```diff
@@ -0,0 +1,117 @@
+const infixToFunction = {
+  "+": (x, y) => x + y,
+  "-": (x, y) => x - y,
+  "*": (x, y) => x * y,
+  "/": (x, y) => x / y
+};
+
+const infixEval = (str, regex) =>
+  str.replace(regex, (_, arg1, fn, arg2) =>
+    infixToFunction[fn](parseFloat(arg1), parseFloat(arg2))
+  );
+
+const highPrecedence = str => {
+  const regex = /([0-9.]+)([*\/])([0-9.]+)/;
+  const str2 = infixEval(str, regex);
+  return str === str2 ? str : highPrecedence(str2);
+};
+
+const isEven = num => num % 2 === 0;
+
+const spreadsheetFunctions = {
+  "": x => x,
+  random: ([x, y]) => Math.floor(Math.random() * y + x),
+  increment: nums => nums.map(x => x + 1),
+  firsttwo: arr => arr.slice(0, 2),
+  lastttwo: arr => arr.slice(-2),
+  even: nums => nums.filter(isEven),
+  sum: nums => nums.reduce((a, x) => a + x),
+  has2: arr => arr.includes(2),
+  nodups: arr => arr.reduce((a, x) => a.includes(x) ? a : a.concat(x), [])
+};
+
+/*
+ES6 introduced a shorthand object literal syntax:
```

Comment:
This will be very useful for learners to know. How about changing the example slightly? It will give us a chance to really show what's going on here. The first lesson could read:

ES6 introduced a shorthand object literal syntax:
```
const a = 10;
const myObject = { a }
console.log(myObject); // { a: 10 }
```
First, move `sum` outside of `spreadsheetFunctions`.
`sum` should be a function expression similar to `isEven`.

The next step could read:

Now reference `sum` inside of `spreadsheetFunctions`.
This both adds it to the functions you can use in the spreadsheet, and allows you to use it throughout your program.